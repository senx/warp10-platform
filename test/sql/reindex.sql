-- This file and its contents are licensed under the Apache License 2.0.
-- Please see the included NOTICE for copyright information and
-- LICENSE-APACHE for a copy of the license.

CREATE TABLE reindex_test(time timestamp, temp float, PRIMARY KEY(time, temp));
CREATE UNIQUE INDEX reindex_test_time_unique_idx ON reindex_test(time);

-- create hypertable with three chunks
SELECT create_hypertable('reindex_test', 'time', chunk_time_interval => 2628000000000);

INSERT INTO reindex_test VALUES ('2017-01-20T09:00:01', 17.5),
                                ('2017-01-21T09:00:01', 19.1),
                                ('2017-04-20T09:00:01', 89.5),
                                ('2017-04-21T09:00:01', 17.1),
                                ('2017-06-20T09:00:01', 18.5),
                                ('2017-06-21T09:00:01', 11.0);

SELECT * FROM test.show_columns('reindex_test');
SELECT * FROM test.show_subtables('reindex_test');

-- show reindexing
REINDEX (VERBOSE) TABLE reindex_test;

\set ON_ERROR_STOP 0
-- this one currently doesn't recurse to chunks and instead gives an
-- error
REINDEX (VERBOSE) INDEX reindex_test_time_unique_idx;
\set ON_ERROR_STOP 1

-- show reindexing on a normal table
CREATE TABLE reindex_norm(time timestamp, temp float);
CREATE UNIQUE INDEX reindex_norm_time_unique_idx ON reindex_norm(time);

INSERT INTO reindex_norm VALUES ('2017-01-20T09:00:01', 17.5),
                                ('2017-01-21T09:00:01', 19.1),
                                ('2017-04-20T09:00:01', 89.5),
                                ('2017-04-21T09:00:01', 17.1),
                                ('2017-06-20T09:00:01', 18.5),
                                ('2017-06-21T09:00:01', 11.0);

REINDEX (VERBOSE) TABLE reindex_norm;
REINDEX (VERBOSE) INDEX reindex_norm_time_unique_idx;

SELECT * FROM test.show_constraintsp('_timescaledb_internal._hyper%');
SELECT * FROM reindex_norm;

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
SELECT chunk_index_clone::regclass::text
FROM _timescaledb_internal.chunk_index_clone('_timescaledb_internal."1_1_reindex_test_pkey"'::regclass);
SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
SELECT * FROM _timescaledb_internal.chunk_index_replace('_timescaledb_internal."1_1_reindex_test_pkey"'::regclass, '_timescaledb_internal."_hyper_1_1_chunk_1_1_reindex_test_pkey"'::regclass);

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
