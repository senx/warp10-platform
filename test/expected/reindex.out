-- This file and its contents are licensed under the Apache License 2.0.
-- Please see the included NOTICE for copyright information and
-- LICENSE-APACHE for a copy of the license.
CREATE TABLE reindex_test(time timestamp, temp float, PRIMARY KEY(time, temp));
CREATE UNIQUE INDEX reindex_test_time_unique_idx ON reindex_test(time);
-- create hypertable with three chunks
SELECT create_hypertable('reindex_test', 'time', chunk_time_interval => 2628000000000);
     create_hypertable     
---------------------------
 (1,public,reindex_test,t)
(1 row)

INSERT INTO reindex_test VALUES ('2017-01-20T09:00:01', 17.5),
                                ('2017-01-21T09:00:01', 19.1),
                                ('2017-04-20T09:00:01', 89.5),
                                ('2017-04-21T09:00:01', 17.1),
                                ('2017-06-20T09:00:01', 18.5),
                                ('2017-06-21T09:00:01', 11.0);
SELECT * FROM test.show_columns('reindex_test');
 Column |            Type             | NotNull 
--------+-----------------------------+---------
 time   | timestamp without time zone | t
 temp   | double precision            | t
(2 rows)

SELECT * FROM test.show_subtables('reindex_test');
                 Child                  | Tablespace 
----------------------------------------+------------
 _timescaledb_internal._hyper_1_1_chunk | 
 _timescaledb_internal._hyper_1_2_chunk | 
 _timescaledb_internal._hyper_1_3_chunk | 
 _timescaledb_internal._hyper_1_4_chunk | 
 _timescaledb_internal._hyper_1_5_chunk | 
(5 rows)

-- show reindexing
REINDEX (VERBOSE) TABLE reindex_test;
INFO:  index "1_1_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_1_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "2_2_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_2_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "3_3_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_3_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "4_4_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_4_chunk_reindex_test_time_unique_idx" was reindexed
INFO:  index "5_5_reindex_test_pkey" was reindexed
INFO:  index "_hyper_1_5_chunk_reindex_test_time_unique_idx" was reindexed
\set ON_ERROR_STOP 0
-- this one currently doesn't recurse to chunks and instead gives an
-- error
REINDEX (VERBOSE) INDEX reindex_test_time_unique_idx;
ERROR:  reindexing of a specific index on a hypertable is unsupported
\set ON_ERROR_STOP 1
-- show reindexing on a normal table
CREATE TABLE reindex_norm(time timestamp, temp float);
CREATE UNIQUE INDEX reindex_norm_time_unique_idx ON reindex_norm(time);
INSERT INTO reindex_norm VALUES ('2017-01-20T09:00:01', 17.5),
                                ('2017-01-21T09:00:01', 19.1),
                                ('2017-04-20T09:00:01', 89.5),
                                ('2017-04-21T09:00:01', 17.1),
                                ('2017-06-20T09:00:01', 18.5),
                                ('2017-06-21T09:00:01', 11.0);
REINDEX (VERBOSE) TABLE reindex_norm;
INFO:  index "reindex_norm_time_unique_idx" was reindexed
REINDEX (VERBOSE) INDEX reindex_norm_time_unique_idx;
INFO:  index "reindex_norm_time_unique_idx" was reindexed
SELECT * FROM test.show_constraintsp('_timescaledb_internal._hyper%');
                 Table                  |      Constraint       | Type |   Columns   |                     Index                     |                                                                     Expr                                                                     | Deferrable | Deferred | Validated 
----------------------------------------+-----------------------+------+-------------+-----------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------+------------+----------+-----------
 _timescaledb_internal._hyper_1_1_chunk | 1_1_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."1_1_reindex_test_pkey" |                                                                                                                                              | f          | f        | t
 _timescaledb_internal._hyper_1_1_chunk | constraint_1          | c    | {time}      | -                                             | (("time" >= 'Thu Jan 19 10:00:00 2017'::timestamp without time zone) AND ("time" < 'Sat Feb 18 20:00:00 2017'::timestamp without time zone)) | f          | f        | t
 _timescaledb_internal._hyper_1_2_chunk | 2_2_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."2_2_reindex_test_pkey" |                                                                                                                                              | f          | f        | t
 _timescaledb_internal._hyper_1_2_chunk | constraint_2          | c    | {time}      | -                                             | (("time" >= 'Tue Mar 21 06:00:00 2017'::timestamp without time zone) AND ("time" < 'Thu Apr 20 16:00:00 2017'::timestamp without time zone)) | f          | f        | t
 _timescaledb_internal._hyper_1_3_chunk | 3_3_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."3_3_reindex_test_pkey" |                                                                                                                                              | f          | f        | t
 _timescaledb_internal._hyper_1_3_chunk | constraint_3          | c    | {time}      | -                                             | (("time" >= 'Thu Apr 20 16:00:00 2017'::timestamp without time zone) AND ("time" < 'Sun May 21 02:00:00 2017'::timestamp without time zone)) | f          | f        | t
 _timescaledb_internal._hyper_1_4_chunk | 4_4_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."4_4_reindex_test_pkey" |                                                                                                                                              | f          | f        | t
 _timescaledb_internal._hyper_1_4_chunk | constraint_4          | c    | {time}      | -                                             | (("time" >= 'Sun May 21 02:00:00 2017'::timestamp without time zone) AND ("time" < 'Tue Jun 20 12:00:00 2017'::timestamp without time zone)) | f          | f        | t
 _timescaledb_internal._hyper_1_5_chunk | 5_5_reindex_test_pkey | p    | {time,temp} | _timescaledb_internal."5_5_reindex_test_pkey" |                                                                                                                                              | f          | f        | t
 _timescaledb_internal._hyper_1_5_chunk | constraint_5          | c    | {time}      | -                                             | (("time" >= 'Tue Jun 20 12:00:00 2017'::timestamp without time zone) AND ("time" < 'Thu Jul 20 22:00:00 2017'::timestamp without time zone)) | f          | f        | t
(10 rows)

SELECT * FROM reindex_norm;
           time           | temp 
--------------------------+------
 Fri Jan 20 09:00:01 2017 | 17.5
 Sat Jan 21 09:00:01 2017 | 19.1
 Thu Apr 20 09:00:01 2017 | 89.5
 Fri Apr 21 09:00:01 2017 | 17.1
 Tue Jun 20 09:00:01 2017 | 18.5
 Wed Jun 21 09:00:01 2017 |   11
(6 rows)

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
                                Index                                |   Columns   | Expr | Unique | Primary | Exclusion | Tablespace 
---------------------------------------------------------------------+-------------+------+--------+---------+-----------+------------
 _timescaledb_internal."1_1_reindex_test_pkey"                       | {time,temp} |      | t      | t       | f         | 
 _timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx | {time}      |      | t      | f       | f         | 
(2 rows)

SELECT chunk_index_clone::regclass::text
FROM _timescaledb_internal.chunk_index_clone('_timescaledb_internal."1_1_reindex_test_pkey"'::regclass);
                      chunk_index_clone                       
--------------------------------------------------------------
 _timescaledb_internal._hyper_1_1_chunk_1_1_reindex_test_pkey
(1 row)

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
                                Index                                |   Columns   | Expr | Unique | Primary | Exclusion | Tablespace 
---------------------------------------------------------------------+-------------+------+--------+---------+-----------+------------
 _timescaledb_internal."1_1_reindex_test_pkey"                       | {time,temp} |      | t      | t       | f         | 
 _timescaledb_internal._hyper_1_1_chunk_1_1_reindex_test_pkey        | {time,temp} |      | t      | t       | f         | 
 _timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx | {time}      |      | t      | f       | f         | 
(3 rows)

SELECT * FROM _timescaledb_internal.chunk_index_replace('_timescaledb_internal."1_1_reindex_test_pkey"'::regclass, '_timescaledb_internal."_hyper_1_1_chunk_1_1_reindex_test_pkey"'::regclass);
 chunk_index_replace 
---------------------
 
(1 row)

SELECT * FROM test.show_indexes('_timescaledb_internal._hyper_1_1_chunk');
                                Index                                |   Columns   | Expr | Unique | Primary | Exclusion | Tablespace 
---------------------------------------------------------------------+-------------+------+--------+---------+-----------+------------
 _timescaledb_internal."1_1_reindex_test_pkey"                       | {time,temp} |      | t      | t       | f         | 
 _timescaledb_internal._hyper_1_1_chunk_reindex_test_time_unique_idx | {time}      |      | t      | f       | f         | 
(2 rows)

